File: 61-rotate-list/rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(head==nullptr or head ->next ==nullptr or k==0){
            return head;
        }
        ListNode * slow (head);
        ListNode * fast(head);
        ListNode * tailLeft=nullptr;
        //tail for left side 
        ListNode * tailRight = nullptr ;
        int len =1;
        //tail of right side after splitting
        while(slow->next!=nullptr){
            slow=slow->next;
            len++;
        }
        k %=len;
        int p =len-k;
        
        // slow=head;
        if(k==0){
            return head;
        }
        slow->next =head;
        while(p--){
                slow = slow ->next;
        }
        // this will be tail after rotating
     
        head =slow->next;
        slow->next = NULL;
        return head;
    }
};

File: 61-rotate-list/rotate-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function rotateRight(head: ListNode | null, k: number): ListNode | null {
    if(head==null) return null ;
    let len: number =0
    let temp : ListNode | null = head;
    let prev : ListNode | null = null;
    let tail : ListNode | null = null;
    while(temp){
        tail=temp;
        temp=temp.next;
        len++;
    }
    k=k%len;
    if(k==0) return head;
    k= len-k;
    temp = head;

    while(k>0){
        prev=temp;
        temp=temp.next;
        k--;
    }
    if(temp==null)return head;
    prev.next=null;
    tail.next=head;

    return temp;
};